From: Marcel Holtmann <marcel@holtmann.org>
Subject: bluetooth: wake up properly after ide timeout expires
Patch-mainline: 2.6.26
References: bnc#390839

Some Bluetooth mice and keyboards do not reconnect after the idle timeout
expired (usually 1800 seconds).
To get them going again, you need to either:
- pull the dongle and plug it again
- switch off the device and on again
- do some dbus-magic to tell userspace to explicitly disconnect the device

In a long and exhausting debugging session with Marcel, we found that the
problem is actually a bug in the hidp kernel module.

I'll attach a patch (from Marcel Holtmann, the bluetooth maintainer) that he
also will push upstream soon, and i'd request that we include the Patch in our
kernel for 11.0.

Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 net/bluetooth/hidp/core.c |   10 ++++++++++
 1 files changed, 10 insertions(+)

--- a/net/bluetooth/hidp/core.c
+++ b/net/bluetooth/hidp/core.c
@@ -581,6 +581,12 @@
 		hid_free_device(session->hid);
 	}
 
+	/* Wakeup user-space polling for socket errors */
+	session->intr_sock->sk->sk_err = EUNATCH;
+	session->ctrl_sock->sk->sk_err = EUNATCH;
+
+	hidp_schedule(session);
+
 	fput(session->intr_sock->file);
 
 	wait_event_timeout(*(ctrl_sk->sk_sleep),
@@ -879,6 +885,10 @@
 			skb_queue_purge(&session->ctrl_transmit);
 			skb_queue_purge(&session->intr_transmit);
 
+			/* Wakeup user-space polling for socket errors */
+			session->intr_sock->sk->sk_err = EUNATCH;
+			session->ctrl_sock->sk->sk_err = EUNATCH;
+
 			/* Kill session thread */
 			atomic_inc(&session->terminate);
 			hidp_schedule(session);
